server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -73,21 +77,27 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<ShardRefreshRequest, ReplicationResponse>> listener
     ) {
         ActionListener.completeWith(listener, () -> {
-            primary.refresh(SOURCE_API);
+            var refreshResult = primary.refresh(SOURCE_API);
             logger.trace("{} refresh request executed on primary", primary.shardId());
-            return new PrimaryResult<>(shardRequest, new ReplicationResponse());
+            // TODO(PS): probably we need to do more to maintain BwC?
+            var shardRefreshRequest = new ShardRefreshRequest(primary.shardId(), refreshResult.getSegmentGeneration());
+            return new PrimaryResult<>(shardRefreshRequest, new ReplicationResponse());
         });
     }
 
     @Override
-    protected void shardOperationOnReplica(BasicReplicationRequest request, IndexShard replica, ActionListener<ReplicaResult> listener) {
-        ActionListener.completeWith(listener, () -> {
-            replica.refresh(SOURCE_API);
-            logger.trace("{} refresh request executed on replica", replica.shardId());
-            return new ReplicaResult();
-        });
+    protected void shardOperationOnReplica(ShardRefreshRequest request, IndexShard replica, ActionListener<ReplicaResult> listener) {
+        if (DiscoveryNode.isStateless(settings)) {
+            replica.waitForSegmentGeneration(request.getSegmentGeneration(), listener.map(gen -> new ReplicaResult()));
```

Comment:
When we have proper shard roles, will we be doing this for all replicas or just search replicas? If the latter, we need to track this follow-up somewhere.

Reply:
In the current state, my assumption is that `shardOperationOnReplica` when in Stateless, means a Search shard. But This is probably not correct once we have indexing replica shards (I guess we will, for availability reasons!?).

Reply:
It's actually already not the right name in `main` because this is called on both replicas and relocating primaries:

https://github.com/elastic/elasticsearch/blob/2f097ba0ab41b04d93770b83c66c744bcc75edae/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java#L202-L206

IOW if you tried to `assert replica.routingEntry().primary() == false` then it'd fail here.

Reply:
> IOW if you tried to assert replica.routingEntry().primary() == false then it'd fail here.

But is this only due to the relocating primary case? Or even when no relocation is in progress that assertion fails?

Reply:
Correct, that's the only time there's two primary shards in `replicationGroup.getReplicationTargets()`.